File: csharp/need-for-speed/NeedForSpeed.cs
using System;
using System.Numerics;


class RemoteControlCar
{

    private int speed = 0;
    private int batteryDrain = 0;
    private int distanceDrived = 0;
    private int battery = 100;

    public RemoteControlCar(int speed, int batteryDrain)
    {
        this.speed = speed;
        this.batteryDrain = batteryDrain;

    }

    public bool BatteryDrained()
        => this.battery < this.batteryDrain;

    public int DistanceDriven()
        => this.distanceDrived;

    public void Drive()
    {
        if(!this.BatteryDrained()) {
            this.distanceDrived += this.speed;
            this.battery -= this.batteryDrain;
        }

    }

    public static RemoteControlCar Nitro()
        => new RemoteControlCar(50, 4);
}

class RaceTrack
{
    int distance = 0;

    public RaceTrack(int distance)
    {
        this.distance = distance;
    }

    public bool TryFinishTrack(RemoteControlCar car)
    {
        while (car.DistanceDriven() < this.distance)
        {
            if (car.BatteryDrained())
                return false;

            car.Drive();
        }

        return true;

    }
}

File: java/need-for-speed/src/main/java/NeedForSpeed.java
class NeedForSpeed {

    private int speed = 0;
    private int batteryDrain = 0;
    private int battery = 100;
    private int drivenDistance = 0;

    public int getSpeed() {
        return this.speed;
    }

    public int getBatteryDrain() {
        return this.batteryDrain;
    }

    public int getBattery() {
        return this.battery;
    }

    public int getDrivenDistance() {
        return this.drivenDistance;
    }

    NeedForSpeed(int speed, int batteryDrain) {
        this.speed = speed;
        this.batteryDrain = batteryDrain;
    }

    public boolean batteryDrained() {
        return this.battery < this.batteryDrain;
    }

    public int distanceDriven() {
        return this.drivenDistance;
    }

    public void drive() {
        if (!this.batteryDrained()) {
            this.drivenDistance += this.speed;
            this.battery -= this.batteryDrain;
        }
    }

    public static NeedForSpeed nitro() {
        return new NeedForSpeed(50, 4);
    }
}

class RaceTrack {

    private int distance = 0;

    RaceTrack(int distance) {
        this.distance = distance;
    }

    public boolean tryFinishTrack(NeedForSpeed car) {
        final double carEfficiency = car.getBattery() / car.getBatteryDrain();
        return carEfficiency * car.getSpeed() >= this.distance;
    }

}
